awx/main/scheduler/task_manager_models.py

```diff
@@ -0,0 +1,120 @@
+# Copyright (c) 2022 Ansible by Red Hat
+# All Rights Reserved.
+import logging
+
+from django.conf import settings
+
+from awx.main.models import (
+    Instance,
+    InstanceGroup,
+)
+
+logger = logging.getLogger('awx.main.scheduler')
+
+
+class TaskManagerInstance:
+    """A class representing minimal data the task manager needs to represent an Instance."""
+
+    def __init__(self, obj, node_type, capacity, hostname):
+        self.obj = obj
+        self.node_type = node_type
+        self.remaining_capacity = capacity
+        self.capacity = capacity
+        self.hostname = hostname
+
+
+class TaskManagerInstances:
+    def __init__(self, active_tasks=[]):
+        instances = Instance.objects.filter(hostname__isnull=False, enabled=True).exclude(node_type='hop').only('node_type', 'capacity', 'hostname', 'enabled')
+        self._remaining_capacity_updated = False
+        instance_list = [
+            TaskManagerInstance(
+                obj=instance,
+                node_type=instance.node_type,
+                capacity=instance.capacity,
+                hostname=instance.hostname,
+            )
+            for instance in instances
+        ]
+        self._update_remaining_capacity(active_tasks)
+        self.instances_by_hostname = {i.hostname: i for i in instance_list}
+
+    def _update_remaining_capacity(self, tasks):
+        """Takes a list of tasks and updates the remaining capacity on our TaskManagerInstance list.
+
+        Computes remaining capacity for all the instances based on currently running and waiting tasks.
+        """
+        if self._remaining_capacity_updated:
+            # Only want to do this once
+            return
+        for task in tasks:
+            if task.status not in ['waiting', 'running']:
+                continue
+            control_instance = self.instances_by_hostname.get(task.controller_node, '')
+            execution_instance = self.instances_by_hostname.get(task.execution_node, '')
+            if execution_instance and execution_instance.node_type in ('hybrid', 'execution'):
+                self.instances_by_hostname[task.execution_node].remaining_capacity -= task.task_impact
+            if control_instance and control_instance.node_type in ('hybrid', 'control'):
+                self.instances_by_hostname[task.controller_node].remaining_capacity -= settings.AWX_CONTROL_NODE_TASK_IMPACT
+
+    def __getitem__(self, hostname):
+        return self.instances_by_hostname.get(hostname)
```

Comment:
It looks like you only use this to get the controlplane IG. If that's true, then I would lean towards axing this because it carries with it risk and awkward questions. Instead you could have a `get_control_plane` method or something.

Reply:
Wait, I was getting confused. At issue is the line

```python
self.controlplane_ig = self.instance_groups.controlplane_ig
```

This doesn't make sense, this would work if you override getattr, but both of these dict-like class (instance list and IG list) define _getitem_. I like the getitem, and use pattern like:

```python
self.instances[control_instance.hostname].obj
```

That looks really good. Still, if you don't use it, I wouldn't put it on `TaskManagerInstanceGroups`, and you still probably need a separate function to get the control plane group.